File: src/Pb2.scala
object Pb2 {
	def solution(directions: List[Tuple]): Int = {
		var depth = 0
		var horizontal = 0
		for((a: String, b: Int) <- directions) {
			a match {
				case "forward" => horizontal += b
				case "down" => depth += b
				case "up" => depth -= b
			}
		}

		return depth * horizontal
	}

	def solution2(directions: List[Tuple]): Int = {
		var depth = 0
		var horizontal = 0
		var aim = 0
		for((a: String, b: Int) <- directions) {
			a match {
				case "forward" => horizontal += b; depth += aim * b
				case "down" => aim += b
				case "up" => aim -= b
			}
		}

		return depth * horizontal
	}	
	
	def main(args: Array[String]) = {
		val directions = io.Source.fromFile("../resources/input2.txt")
			.getLines.toList.map(line => line.split(" ")).map(x => (x(0), x(1).toInt))
		println("Solution part n° 1: " + solution(directions))
		println("Solution part n° 2: " + solution2(directions))
	}	
}

File: src/pb2.py
from typing import Tuple
from tools import print_solution, read_file


def solution(directions: list[Tuple[str, int]]) -> int:
	depth = 0
	horizontal = 0

	for direction, value in directions:
		if direction == 'forward':
			horizontal += value
		elif direction == 'down':
			depth += value
		elif direction == 'up':
			depth -= value
	
	return depth * horizontal


def solution2(directions: list[Tuple[str, int]]) -> int:
	depth = 0
	horizontal = 0
	aim = 0

	for direction, value in directions:
		if direction == 'forward':
			horizontal += value
			depth += aim * value
		elif direction == 'down':
			aim += value
		elif direction == 'up':
			aim -= value
	
	return depth * horizontal


if __name__ == '__main__':
	directions = [(dir, int(val)) 
		for dir, val in [line.replace('\n', '').split(' ')  
			for line in read_file('../resources/input2.txt')]]

	print_solution(1, solution(directions))
	print_solution(2, solution2(directions))